core/src/services/alluxio/core.rs

```diff
@@ -0,0 +1,407 @@
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+use std::fmt::Debug;
+use std::fmt::Formatter;
+
+use http::Request;
+
+use http::header::RANGE;
+use http::Response;
+
+use serde::{Deserialize, Serialize};
+
+use crate::raw::*;
+use crate::*;
+
+#[derive(Debug, Serialize)]
+#[serde(untagged)]
+enum Op {
+    CreateFile {
+        /// Whether to create file recursively
+        #[serde(skip_serializing_if = "Option::is_none")]
+        recursive: Option<bool>,
+    },
+    CreateDir {
+        /// Whether to create dir recursively
+        #[serde(skip_serializing_if = "Option::is_none")]
+        recursive: Option<bool>,
+    },
+    OpenFile,
+    Read,
+    Write,
+    Close,
+    Delete,
+    Rename,
+    ListStatus,
+    GetStatus,
+}
+
+/// Metadata of alluxio object
+#[derive(Debug, Deserialize)]
+#[serde(rename_all = "camelCase")]
+pub(super) struct FileInfo {
+    /// The path of the object
+    pub path: String,
+    /// The last modification time of the object
+    pub last_modification_time_ms: i64,
+    /// Whether the object is a folder
+    pub folder: bool,
+    /// The length of the object in bytes
+    pub length: u64,
+}
+
+impl TryFrom<FileInfo> for Metadata {
+    type Error = Error;
+
+    fn try_from(file_info: FileInfo) -> Result<Metadata> {
+        let mut metadata = if file_info.folder {
+            Metadata::new(EntryMode::DIR)
+        } else {
+            Metadata::new(EntryMode::FILE)
+        };
+        metadata
+            .set_content_length(file_info.length)
+            .set_last_modified(parse_datetime_from_from_timestamp_millis(
+                file_info.last_modification_time_ms,
+            )?);
+        Ok(metadata)
+    }
+}
+
+impl Op {
+    /// Get operation name
+    fn as_str(&self) -> &str {
+        match self {
+            Op::CreateFile { .. } => "create-file",
+            Op::CreateDir { .. } => "create-dir",
+            Op::OpenFile => "open-file",
+            Op::Read => "read",
+            Op::Write => "write",
+            Op::Close => "close",
+            Op::Delete => "delete",
+            Op::Rename => "rename",
+            Op::ListStatus => "list-status",
+            Op::GetStatus => "get-status",
+        }
+    }
+}
+
+/// the status code of alluxio
+#[derive(Debug, Deserialize)]
+#[serde(rename_all = "SCREAMING_SNAKE_CASE")]
+enum StatusCode {
+    AlreadyExists,
+    NotFound,
+    Internal,
+}
+
+/// the error response of alluxio
+#[derive(Debug, Deserialize)]
+#[serde(rename_all = "camelCase")]
+struct ErrRsponse {
+    status_code: StatusCode,
+    message: String,
+}
+
+impl From<ErrRsponse> for Error {
```

Comment:
Please don't impl `From<ErrRsponse> for Error` which could leak error unexpectedly. Please use function like `parse_error_response(ErrResponse) -> Error`.